File: bot.js
const TelegramBot = require('node-telegram-bot-api');
const sqlite3 = require('sqlite3').verbose();

// Токен вашего бота, полученный через BotFather
const token = '8075874421:AAHwWSia-Hs3bHeTFrTVPJnOVVofYBYPD1o';
const bot = new TelegramBot(token, { polling: true });

// Создание базы данных SQLite (или подключение к существующей)
const db = new sqlite3.Database('./calculations.db');

// Создание таблицы для хранения расчетов, если таблица еще не существует
db.run(`
  CREATE TABLE IF NOT EXISTS calculations (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    propertyType TEXT,
    repairType TEXT,
    rooms INTEGER,
    bathrooms INTEGER,
    area REAL,
    contactMethod TEXT,
    chatId INTEGER,
    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
  );
`);

// Структура данных для отслеживания состояния
let userData = {};

// Стартовое сообщение и кнопка
bot.onText(/\/start/, (msg) => {
  const chatId = msg.chat.id;
  userData[chatId] = {}; // Инициализация данных для пользователя
  
  const options = {
  reply_markup: {
    keyboard: [
      [{ text: "Расчет стоимости ремонта" }],
      [{ text: "Отзывы", url: 'https://vk.com/topic-214261496_48716771' }],
	  [{ text: "Топ часто задаваемых вопросов"}]
    ],
    resize_keyboard: true,
	one_time_keyboard: true
  }
};

bot.on('message', (msg) => {
  const chatId = msg.chat.id;

  if (msg.text === 'Отзывы') {
    // Отправляем сообщение с текстом ссылки
    bot.sendMessage(chatId, 'Нажмите на ссылку, чтобы открыть отзывы: https://vk.com/topic-214261496_48716771');
  }
});

bot.sendMessage(chatId, 'Добро пожаловать в ПроРемонт! Я с радостью помогу вам разобраться во всех тонкостях ремонта и выбрать оптимальные решения!', options);
});

// Обработка нажатия на кнопку "Расчет стоимости ремонта"
bot.onText(/Расчет стоимости ремонта/, (msg) => {
  const chatId = msg.chat.id;

  const options = {
    reply_markup: {
      keyboard: [
        [{ text: 'Новостройка' }, { text: 'Вторичка' }],
        [{ text: 'Частный дом' }, { text: 'Коммерческое' }]
      ],
      resize_keyboard: true
    }
  };

  bot.sendMessage(chatId, 'Выберите тип недвижимости:', options);
});

// Обработка выбора типа недвижимости
bot.onText(/^(Новостройка|Вторичка|Частный дом|Коммерческое)$/, (msg, match) => {
  const chatId = msg.chat.id;
  const type = match[0];

  userData[chatId].propertyType = type; // Сохранение выбранного типа недвижимости

  const options = {
    reply_markup: {
      keyboard: [
        [{ text: 'Косметический' }, { text: 'Капитальный' }],
        [{ text: 'Дизайнерский' }]
      ],
      resize_keyboard: true
    }
  };

  bot.sendMessage(chatId, 'Выберите вид ремонта:', options);
});

// Обработка выбора вида ремонта
bot.onText(/^(Косметический|Капитальный|Дизайнерский)$/, (msg, match) => {
  const chatId = msg.chat.id;
  const repairType = match[0];

  userData[chatId].repairType = repairType; // Сохранение выбранного типа ремонта

  const options = {
    reply_markup: {
      keyboard: [
        [{ text: '1' }, { text: '2' }, { text: '3' }],
        [{ text: '4 или более' }]
      ],
      resize_keyboard: true
    }
  };

  bot.sendMessage(chatId, 'Количество комнат:', options);
});

// Обработка выбора количества комнат
bot.onText(/^(1|2|3|4 или более)$/, (msg, match) => {
  const chatId = msg.chat.id;
  const rooms = match[0];

  userData[chatId].rooms = rooms; // Сохранение количества комнат

  const options = {
    reply_markup: {
      keyboard: [
        [{ text: '1' }, { text: '2' }, { text: '3' }]
      ],
      resize_keyboard: true
    }
  };

  bot.sendMessage(chatId, 'Количество санузлов:', options);
});

// Обработка выбора количества санузлов
bot.onText(/^(1|2|3)$/, (msg, match) => {
  const chatId = msg.chat.id;
  const bathrooms = match[0];

  userData[chatId].bathrooms = bathrooms; // Сохранение количества санузлов

  bot.sendMessage(chatId, 'Укажите площадь квартиры (в м²):');
});

// Обработка площади квартиры
bot.on('message', (msg) => {
  const chatId = msg.chat.id;
  const text = msg.text;

  if (userData[chatId] && userData[chatId].bathrooms && !isNaN(parseFloat(text))) {
    // Площадь квартиры
    userData[chatId].area = text;

    const options = {
      reply_markup: {
        keyboard: [
          [{ text: 'Телефон мастеру' }, { text: 'WhatsApp' }, { text: 'Telegram' }]
        ],
        resize_keyboard: true
      }
    };

    bot.sendMessage(chatId, 'Куда отправить расчет стоимости?', options);
  }
});

// Обработка выбора канала для отправки данных
bot.onText(/^(Телефон мастеру|WhatsApp|Telegram)$/, (msg, match) => {
  const chatId = msg.chat.id;
  const contactMethod = match[0];

  userData[chatId].contactMethod = contactMethod; // Сохранение канала связи

  const summary = `
    Ремонт: ${userData[chatId].repairType}
    Тип недвижимости: ${userData[chatId].propertyType}
    Количество комнат: ${userData[chatId].rooms}
    Количество санузлов: ${userData[chatId].bathrooms}
    Площадь квартиры: ${userData[chatId].area} м²
    Канал связи: ${userData[chatId].contactMethod}
  `;

  // Сохранение данных в базу данных SQLite
  db.run(`
    INSERT INTO calculations (propertyType, repairType, rooms, bathrooms, area, contactMethod, chatId)
    VALUES (?, ?, ?, ?, ?, ?, ?)`,
    [userData[chatId].propertyType, userData[chatId].repairType, userData[chatId].rooms, userData[chatId].bathrooms, userData[chatId].area, userData[chatId].contactMethod, chatId],
    function(err) {
      if (err) {
        bot.sendMessage(chatId, 'Произошла ошибка при сохранении данных.');
      } else {
        bot.sendMessage(chatId, 'Ваш расчет успешно сохранен.');
      }
    }
  );

  // Генерация ссылки для выбранного канала связи
  let sendLink = '';
  if (contactMethod === 'Телефон мастеру') {
    sendLink = `Телефон: +7 (919) 916-20-49`; // Подставьте телефон мастера (например, +1234567890)
  } else if (contactMethod === 'WhatsApp') {
    sendLink = `https://wa.me/+79199162049?text=${encodeURIComponent(summary)}`; // Ссылка на WhatsApp
  } else if (contactMethod === 'Telegram') {
    sendLink = `https://t.me/username?text=${encodeURIComponent(summary)}`; // Ссылка на Telegram (поменяйте на никнейм мастера)
  }

  // Отправка ссылки на выбранный канал
  bot.sendMessage(chatId, `Ваш расчет отправлен на: ${contactMethod}`);

  // Отправка ссылки мастеру
  bot.sendMessage(chatId, `Вот ссылка для связи с мастером: ${sendLink}`);

  // Кнопка для возврата на стартовую страницу
  const options = {
    reply_markup: {
      keyboard: [
        [{ text: "Начать заново" }]
      ],
      resize_keyboard: true
    }
  };
  bot.sendMessage(chatId, 'Вы можете начать новый расчет или /start', options);
});

// Обработка нажатия на кнопку "Начать заново"
bot.onText(/Начать заново/, (msg) => {
  const chatId = msg.chat.id;
  userData[chatId] = {}; // Сброс данных пользователя

  const options = {
    reply_markup: {
      keyboard: [
        [{ text: "Расчет стоимости ремонта" }]
		[{ text: "Отзывы", url: 'https://vk.com/topic-214261496_48716771' }],
		[{ text: "Топ часто задаваемых вопросов"}]
      ],
      resize_keyboard: true
    }
  };
  
bot.sendMessage(chatId, 'Добро пожаловать в ПроРемонт! Я с радостью помогу вам разобраться во всех тонкостях ремонта и выбрать оптимальные решения!', options);
});

File: Резервная копия/bot.py
import logging
import sqlite3
from aiogram import Bot, Dispatcher, types
from aiogram.types import ParseMode
from aiogram.utils import executor
import os
from aiogram.contrib.middlewares.logging import LoggingMiddleware
from aiogram.types import InlineKeyboardMarkup, InlineKeyboardButton

# Токен вашего бота, полученный через BotFather
API_TOKEN = '8075874421:AAHwWSia-Hs3bHeTFrTVPJnOVVofYBYPD1o'

# Настройка логирования
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Инициализация бота и диспетчера
bot = Bot(token=API_TOKEN)
dp = Dispatcher(bot)
dp.middleware.setup(LoggingMiddleware())

# Создание базы данных SQLite (или подключение к существующей)
db = sqlite3.connect('./DB/calculations.db')
cursor = db.cursor()

dbBook = sqlite3.connect('./DB/tour_bookings.db')
cursorBook = dbBook.cursor()

# Создание таблиц, если их еще нет
cursor.execute('''CREATE TABLE IF NOT EXISTS calculations (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    propertyType TEXT,
    repairType TEXT,
    rooms INTEGER,
    bathrooms INTEGER,
    area REAL,
    contactMethod TEXT,
    chatId INTEGER,
    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
);''')

cursorBook.execute('''CREATE TABLE IF NOT EXISTS bookings (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    name TEXT,
    phone TEXT,
    date TEXT,
    time TEXT
);''')

# Структура данных для отслеживания состояния пользователя
userData = {}

# Команда /start
@dp.message_handler(commands=['start'])
async def start(message: types.Message):
    chat_id = message.chat.id
    userData[chat_id] = {}

    keyboard = InlineKeyboardMarkup(row_width=2)
    keyboard.add(
        InlineKeyboardButton("🏠Расчет стоимости ремонта", callback_data='calculate_cost'),
        InlineKeyboardButton('📸 Портфолио наших объектов', url='https://proremont18.ru/nashi-rabotyi/'),
        InlineKeyboardButton("📝Отзывы", url='https://vk.com/app6326142_-214261496'),
        InlineKeyboardButton("📝Записаться на экскурсию", callback_data='book_tour'),
        InlineKeyboardButton("❓Топ часто задаваемых вопросов(разработка)", callback_data='faq'),
        InlineKeyboardButton("🎁 ПОДАРОК", callback_data='gift'),
        InlineKeyboardButton('Контактная информация 📞', callback_data='contact_info'),
        InlineKeyboardButton('👨‍💼 Связаться с менеджером', callback_data='contact_manager')
    )
    await message.answer(
        '🛠Добро пожаловать в ПроРемонт! Наш бот может рассчитать стоимость ремонта, записать вас на экскурсию по нашим действующим объектам, а также вы можете посмотреть портфолио наших объектов и отзывы заказчиков!🛠',
        reply_markup=keyboard
    )

# Обработка нажатий на инлайн-кнопки
@dp.callback_query_handler(lambda c: c.data)
async def process_callback(callback_query: types.CallbackQuery):
    chat_id = callback_query.message.chat.id
    data = callback_query.data

    if data == 'book_tour':
        await bot.send_message(chat_id, 'Введите ваше имя:')
        userData[chat_id] = {}  # Инициализация данных пользователя
    elif data == 'contact_info':
        contact_info = """
        Контактная информация:
        📍 Адрес: г. Ижевск, ул. Металлургов, 2
        📞 Телефон: +7 (919) 916-20- 49
        ✉️ E-mail: proremont.18@yandex.ru
        🕒 Режим работы: Пн. – Пт. с 8:00 до 17:00
        """
        await bot.send_message(chat_id, contact_info)
    elif data == 'contact_manager':
        contact_info = """
        👨‍💼 *Связаться с нашим менеджером:*
        📞 Телефон: +7 (919) 916-20-49
        📧 E-mail: proremont.18@yandex.ru
        """
        await bot.send_message(chat_id, contact_info)
    elif data == 'gift':
        await send_gift_files(chat_id)
    elif data == 'calculate_cost':
        await bot.send_message(chat_id, 'Выберите тип недвижимости:', reply_markup=types.ReplyKeyboardMarkup(
            inline_keyboard=[
                [InlineKeyboardButton('Новостройка', callback_data='new_building'), InlineKeyboardButton('Вторичка', callback_data='secondary')],
                [InlineKeyboardButton('Частный дом', callback_data='house'), InlineKeyboardButton('Коммерческое', callback_data='commercial')]
            ]
        ))

# Функция отправки подарков
async def send_gift_files(chat_id):
    loading_msg = await bot.send_message(chat_id, 'Загрузка файлов...')
    pdf_files = ['pdf/Aq.pdf', 'pdf/compressed.pdf']

    for file in pdf_files:
        try:
            await bot.send_document(chat_id, open(file, 'rb'))
            logger.info(f'Файл {file} отправлен успешно.')
        except Exception as e:
            logger.error(f'Ошибка при отправке файла {file}: {e}')
            await bot.send_message(chat_id, 'Ошибка при отправке файла.')

    await bot.delete_message(chat_id, loading_msg.message_id)
    await bot.send_message(chat_id, 'Файлы успешно отправлены!')

# Обработка сообщений
@dp.message_handler(content_types=['text'])
async def handle_message(message: types.Message):
    chat_id = message.chat.id
    text = message.text

    if chat_id in userData:
        if 'name' not in userData[chat_id]:
            userData[chat_id]['name'] = text
            await bot.send_message(chat_id, 'Введите ваш номер телефона:')
        elif 'phone' not in userData[chat_id]:
            userData[chat_id]['phone'] = text
            await bot.send_message(chat_id, 'Выберите дату (например, 2024-11-13):')
        elif 'date' not in userData[chat_id]:
            userData[chat_id]['date'] = text
            await bot.send_message(chat_id, 'Укажите время экскурсии (например, 15:00):')
        elif 'time' not in userData[chat_id]:
            userData[chat_id]['time'] = text
            await bot.send_message(chat_id, 'Спасибо за запись! Мы с вами свяжемся.')

            # Сохранение в базе данных
            name = userData[chat_id]['name']
            phone = userData[chat_id]['phone']
            date = userData[chat_id]['date']
            time = userData[chat_id]['time']

            cursorBook.execute('INSERT INTO bookings (name, phone, date, time) VALUES (?, ?, ?, ?)',
                               (name, phone, date, time))
            dbBook.commit()

            # Уведомление администратора
            admin_chat_id = '1426960007'  # Замените на ID администратора
            await bot.send_message(admin_chat_id, f'Новая запись на экскурсию: \nИмя: {name}\nТелефон: {phone}\nДата: {date}\nВремя: {time}')

            # Очистка данных
            del userData[chat_id]
        else:
            await bot.send_message(chat_id, 'Я вас не понимаю.')

if __name__ == '__main__':
    executor.start_polling(dp, skip_updates=True)

@dp.callback_query_handler(lambda c: c.data == 'calculate_cost')
async def handle_calculate_cost(callback_query: types.CallbackQuery):
    chat_id = callback_query.message.chat.id
    
    # Создание инлайн клавиатуры для выбора типа недвижимости
    keyboard = InlineKeyboardMarkup(row_width=2)
    keyboard.add(
        InlineKeyboardButton('Новостройка', callback_data='new_building'),
        InlineKeyboardButton('Вторичка', callback_data='secondary'),
        InlineKeyboardButton('Частный дом', callback_data='house'),
        InlineKeyboardButton('Коммерческое', callback_data='commercial')
    )
    
    await bot.send_message(
        chat_id,
        'Выберите тип недвижимости:',
        reply_markup=keyboard
    )